File: backend/internal/domain/notebook.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

type NoteTag string

const (
	NoteTagStrange    NoteTag = "strange"
	NoteTagSuspicious NoteTag = "suspicious"
	NoteTagLie        NoteTag = "lie"
	NoteTagKey        NoteTag = "key"
)

type NotebookEntryType string

const (
	NotebookEntryTypeStatement          NotebookEntryType = "statement"
	NotebookEntryTypeAnalysis           NotebookEntryType = "analysis"
	NotebookEntryTypeAlibiCheck         NotebookEntryType = "alibi_check"
	NotebookEntryTypeCameraRequest      NotebookEntryType = "camera_request"
	NotebookEntryTypeTransactionRequest NotebookEntryType = "transaction_request"
)

type NotebookEntry struct {
	ID          uuid.UUID         `bson:"id"`
	Type        NotebookEntryType `bson:"type"`
	CharacterID *uuid.UUID        `bson:"character_id,omitempty"`
	Description string            `bson:"description"`
	UserTags    []NoteTag         `bson:"user_tags"`
	UserNote    *string           `bson:"user_note,omitempty"`
	Timestamp   time.Time         `bson:"timestamp"`
}

type ChronologyEventType string

const (
	ChronologyEventTypeCaseStarted      ChronologyEventType = "case_started"
	ChronologyEventTypeInterrogation    ChronologyEventType = "interrogation"
	ChronologyEventTypeLabAnalysis      ChronologyEventType = "lab_analysis"
	ChronologyEventTypeAlibiCheck       ChronologyEventType = "alibi_check"
	ChronologyEventTypeCameraReview     ChronologyEventType = "camera_review"
	ChronologyEventTypeTransactionCheck ChronologyEventType = "transaction_check"
	ChronologyEventTypeFinalReport      ChronologyEventType = "final_report"
)

func ChronologyEventTypeFromAction(action ActionType) ChronologyEventType {
	switch action {
	case ActionTypeAlibiCheck:
		return ChronologyEventTypeAlibiCheck
	case ActionTypeCameraReview:
		return ChronologyEventTypeCameraReview
	case ActionTypeTransactionCheck:
		return ChronologyEventTypeTransactionCheck
	default:
		return ChronologyEventTypeLabAnalysis
	}
}

type ChronologyEntry struct {
	ID        uuid.UUID           `bson:"id"`
	SessionID uuid.UUID           `bson:"session_id"`
	EventType ChronologyEventType `bson:"event_type"`
	OriginID  *uuid.UUID          `bson:"origin_id,omitempty"`
	Title     Translation         `bson:"title_translation"`
	Timestamp time.Time           `bson:"timestamp"`
	Details   []NotebookEntry     `bson:"details"`
}

func NewChronologyEntry(eventType ChronologyEventType, originID *uuid.UUID, title Translation, timestamp time.Time) *ChronologyEntry {
	return &ChronologyEntry{
		ID:        uuid.New(),
		EventType: eventType,
		OriginID:  originID,
		Title:     title,
		Timestamp: timestamp,
		Details:   []NotebookEntry{},
	}
}

type ActionType string

const (
	ActionTypeDNAAnalysis      ActionType = "dna_analysis"
	ActionTypeFingerprints     ActionType = "fingerprints"
	ActionTypeAlibiCheck       ActionType = "alibi_check"
	ActionTypeCameraReview     ActionType = "camera_review"
	ActionTypeCallHistory      ActionType = "call_history"
	ActionTypeTransactionCheck ActionType = "transaction_check"
)

func (t ActionType) Cost() int {
	switch t {
	case ActionTypeDNAAnalysis, ActionTypeFingerprints, ActionTypeAlibiCheck:
		return 1
	default:
		return 2
	}
}

type ActionReport struct {
	ID          uuid.UUID   `bson:"_id"`
	SessionID   uuid.UUID   `bson:"session_id"`
	Type        ActionType  `bson:"type"`
	Title       Translation `bson:"title_translation"`
	Body        string      `bson:"body"`
	EvidenceID  *uuid.UUID  `bson:"evidence_id,omitempty"`
	CharacterID *uuid.UUID  `bson:"character_id,omitempty"`
	Timestamp   time.Time   `bson:"timestamp"`
}

File: mobile/lib/models/notebook.dart
enum NoteTag { strange, suspicious, lie, key }

enum NotebookEntryType { statement, analysis, alibiCheck, cameraRequest, transactionRequest }

class NotebookEntry {
  final String id;
  final NotebookEntryType type;
  final String? characterId;
  final String description;
  final List<NoteTag> userTags;
  final String? userNote;
  final DateTime timestamp;

  const NotebookEntry({
    required this.id,
    required this.type,
    this.characterId,
    required this.description,
    this.userTags = const [],
    this.userNote,
    required this.timestamp,
  });

  String get typeLabel {
    switch (type) {
      case NotebookEntryType.statement:
        return 'показания';
      case NotebookEntryType.analysis:
        return 'анализ';
      case NotebookEntryType.alibiCheck:
        return 'проверка алиби';
      case NotebookEntryType.cameraRequest:
        return 'запись с камер';
      case NotebookEntryType.transactionRequest:
        return 'транзакция';
    }
  }

  String get tagsLabel {
    if (userTags.isEmpty) return '';
    return userTags
        .map((t) {
          switch (t) {
            case NoteTag.strange:
              return 'странно';
            case NoteTag.suspicious:
              return 'подозрительно';
            case NoteTag.lie:
              return 'ложь';
            case NoteTag.key:
              return 'ключевое';
          }
        })
        .join(', ');
  }

  NotebookEntry copyWith({
    String? id,
    NotebookEntryType? type,
    String? characterId,
    String? description,
    List<NoteTag>? userTags,
    String? userNote,
    bool clearTags = false,
    bool clearNote = false,
    DateTime? timestamp,
  }) {
    return NotebookEntry(
      id: id ?? this.id,
      type: type ?? this.type,
      characterId: characterId ?? this.characterId,
      description: description ?? this.description,
      userTags: clearTags ? [] : userTags ?? this.userTags,
      userNote: clearNote ? null : userNote ?? this.userNote,
      timestamp: timestamp ?? this.timestamp,
    );
  }

  static NotebookEntryType entryTypeFromString(String value) {
    switch (value) {
      case 'analysis':
        return NotebookEntryType.analysis;
      case 'alibi_check':
        return NotebookEntryType.alibiCheck;
      case 'camera_request':
        return NotebookEntryType.cameraRequest;
      case 'transaction_request':
        return NotebookEntryType.transactionRequest;
      default:
        return NotebookEntryType.statement;
    }
  }

  factory NotebookEntry.fromJson(Map<String, dynamic> json) {
    return NotebookEntry(
      id: json['id'] as String,
      type: entryTypeFromString(json['type'] as String),
      characterId: json['character_id'] as String?,
      description: json['description'] as String,
      userTags:
          json['user_tags'] != null
              ? (json['user_tags'] as List<dynamic>).map((e) => NoteTag.values.byName(e as String)).toList()
              : [],
      userNote: json['user_note'] as String?,
      timestamp: DateTime.parse(json['timestamp'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'type': type.name,
      'character_id': characterId,
      'description': description,
      'user_tags': userTags.map((t) => t.name).toList(),
      'user_note': userNote,
      'timestamp': timestamp.toIso8601String(),
    };
  }
}
